Grasshopper_UI/Templates/BakeableParam.cs

```diff
@@ -0,0 +1,101 @@
+ * The BHoM is free software: you can redistribute it and/or modify         
+ * it under the terms of the GNU Lesser General Public License as published by  
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.Engine.Grasshopper;
+using BH.UI.Grasshopper.Properties;
+using BH.UI.Grasshopper.Templates;
+using Grasshopper.Kernel;
+using System;
+using Rhino;
+using Rhino.DocObjects;
+using System.Collections.Generic;
+using Grasshopper.Kernel.Types;
+
+namespace BH.UI.Grasshopper.Templates
+{
+    public abstract class BakeableParam<T> : BHoMParam<T>, IGH_BakeAwareObject where T : class, IGH_Goo
+    {
+        /*******************************************/
+        /**** Properties                        ****/
+        /*******************************************/
+
+        public override bool IsPreviewCapable
+        {
+            get
+            {
+                if (VolatileDataCount < m_MaxItemsPreview || m_ForcePreview && !Hidden)
+                {
+                    // Setting IsPreviewCapable from true to false clears the grasshopper geometry cache
+                    // be mindful to use m_ForcePreview and Hidden variables only in case the number of objects
+                    // to display is actually high
+                    this.ClearRuntimeMessages();
+                    return base.IsPreviewCapable;
+                }
+                Engine.Reflection.Compute.RecordNote("Preview has been disabled to prevent a slowdown due to the high number of objects." +
+                    "Right click and set the a new items limit to force the preview at your own risk.");
+                Helpers.ShowEvents(this, Engine.Reflection.Query.CurrentEvents());
+                return false;
+            }
+        }
+
+        public bool IsBakeCapable { get; } = true;
+
+
+        /***************************************************/
+        /**** IGH_BakeAwareObject methods               ****/
+        /***************************************************/
+
+        public void BakeGeometry(RhinoDoc doc, List<Guid> obj_ids)
+        {
+            foreach (Engine.Grasshopper.Objects.GH_BakeableObject<T> item in this.VolatileData.AllData(true))
```

Comment:
```suggestion
            foreach (T item in this.VolatileData.AllData(true))
```

The T on the param level here is already a Goo, so should be just T, not `....Goo<T>`

It required the additional constraint above as well to work.

Layercake GH.....

Reply:
Instead of changing to T could instead had changed to `IGH_BakeAwareData` I think

Reply:
> Instead of changing to T could instead had changed to `IGH_BakeAwareData` I think, but can leave it up to you @adecler

Yes, this is more correct as the interface `IGH_Goo` (which is the constraint on `T`) doesn't provide a `BakeGeometry` method

Reply:
Yeah, the the `T` solution would require the additional constraint on the `T` as shown above :)

Think we probably should put that in anyway, as if we want the Param to be bakeable, then the Goo should be as well (whether we use the `T` or the explicit interface solution matters less I think)